### PR TITLE
Fix performance regression in systhreads on Windows

### PR DESCRIPTION
**Pull request by dra27**
The bug - located with with @patricoferris's assistance - is that the tick thread in systhreads was using `select` in order to suspend the tick thread. The Windows version of `select`, does not allow all three fdsets to be empty, so it was returning an error and the tick thread constantly calling `caml_thread_yield` with disastrous performance results!

The fix is simple - the `Sleep` function should be used instead, which is what OCaml 4.x does. This PR goes further, because `otherlibs/systhreads/st_stubs.c` is really getting peppered with `#ifdef _WIN32`. At present, this PR is a request for comment (especially from @xavierleroy, cf. #10977): it's best viewed commit by commit. I've suggested creating a new `st_pthreads.h` header which can be shared between both `st_win32.h` and `st_posix.h`.

The commits can be separated/rejected as wanted, but they conflict with each other so it might be easy to consider the changes initially in just the one PR. It is of course possible to fix the bug without the refactor.

Note that, contrary to my opinion in the last core developer meeting, I'm not expecting that we'll choose to reimplement st in terms of Windows primitives any time soon: winpthreads really is very easy to compile with MSVC (there are two minor patches which might be worth trying to upstream to the mingw-w64 project) and its licence is MIT. I think the only reason to switch the native Windows ports to using direct Windows API calls again would be because we came up with a faster direct implementation (and it's not obvious that we would).

### CODE REVIEW

File: otherlibs/systhreads/st_posix.h [hunk 1]
```diff
@@ -15,331 +15,22 @@
...
-  if (rc != 0) { caml_stat_free(e); return rc; }
-  rc = pthread_cond_init(&e->triggered, NULL);
-  if (rc != 0)
-  { pthread_mutex_destroy(&e->lock); caml_stat_free(e); return rc; }
-  e->status = 0;
-  *res = e;
-  return 0;
+  struct timeval timeout = {0, msec * 1000};
+  /* select() seems to be the most efficient way to suspend the
+     thread for sub-second intervals */
```
Comment (xavierleroy) [inline otherlibs/systhreads/st_posix.h]: This comment dates back to the 1990's.  In this century, `nanosleep` would be just as good.  Let's remove the comment?

File: otherlibs/systhreads/st_pthreads.h [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+   included in case OCaml is used for an Android CLI utility. */
+int pthread_atfork(void (*prepare)(void), void (*parent)(void),
+                   void (*child)(void));
+#endif
+
+static int st_atfork(void (*fn)(void))
+{
+  caml_atfork_hook = fn;
+  return 0;
+}
```
Comment (xavierleroy) [inline otherlibs/systhreads/st_pthreads.h]: The Android declaration is useless now that we don't call `pthread_atfork` and should be removed.

 `st_atfork` could also be removed, just by inlining `caml_atfork_hook = caml_thread_reinitialize;` in st_stubs.c, function `caml_thread_initialize`.

File: otherlibs/systhreads/st_stubs.c [hunk 3]
```diff
@@ -34,6 +34,8 @@
 #include "caml/sys.h"
 #include "caml/memprof.h"
 
+#include "../../runtime/sync_posix.h"
```
Comment (xavierleroy) [inline otherlibs/systhreads/st_stubs.c:37]: Why is this needed?  If I comment it out it still compiles without warnings...

**Review by xavierleroy — COMMENTED** [review]
That's quite the rabbit hole you fell into while fixing the `select` performance bug!

I like the idea of using the pthread API throughout the OCaml runtime system, with the help of pthread-win32 for the Windows ports.  Indeed, runtime/domains.c relies a lot on pthread, and making it POSIX/Win32 agnostic would be a nightmare.  

Note however the comment in our CI scripts:
```
     # This is needed until the winpthreads support is implemented directly
    export PATH="$PATH:/usr/x86_64-w64-mingw32/sys-root/mingw/bin"
```
It's been my experience that if the pthread-win32 DLL is nowhere in the path, OCaml executables just crash without any explanation.  Will our Windows users understand this?  (They are used to suffer already...)  Could pthread-win32 be linked statically instead?

A few remarks below.  In particular, there's an opportunity to simplify "at fork" processing even more.

File: otherlibs/systhreads/st_stubs.c (hunk 3 above)
Comment (dra27) [reply otherlibs/systhreads/st_stubs.c:37]: Oops, `sync_posix.h` should have been included in `st_posix.h` (where it's used), not `st_pthreads.h` (where it's not).

I've pushed a correcting commit, but perhaps this is silly - `st_pthreads.h` already relies on OCaml runtime functions (i.e. it's not a pure implementation of st anymore), so an alternative would be to include `sync_posix.h` here _in `st_stubs.c`_, and not include it in any of the header files (just like the `caml/` includes) which would also remove the need for the change in `runtime/sync_posix.h`.

File: otherlibs/systhreads/st_posix.h (hunk 1 above)
Comment (dra27) [reply otherlibs/systhreads/st_posix.h]: I've pushed a version using `nanosleep`, but given that it involves two versions of the code, I'd be quite happy to leave that until we return to #10505... I find the (necessary) re-initialisation of the timeval struct on each tick with select very irritating, however unimportant that may be!

File: otherlibs/systhreads/st_pthreads.h (hunk 2 above)
Comment (dra27) [reply otherlibs/systhreads/st_pthreads.h]: Done!

**Comment by dra27** [comment]
> That's quite the rabbit hole you fell into while fixing the `select` performance bug!

I feel increasing guilt with every `#ifdef _WIN32` I add to our code...!

> I like the idea of using the pthread API throughout the OCaml runtime system, with the help of pthread-win32 for the Windows ports. Indeed, runtime/domains.c relies a lot on pthread, and making it POSIX/Win32 agnostic would be a nightmare.
> 
> Note however the comment in our CI scripts:
> 
> ```
>      # This is needed until the winpthreads support is implemented directly
>     export PATH="$PATH:/usr/x86_64-w64-mingw32/sys-root/mingw/bin"
> ```
> 
> It's been my experience that if the pthread-win32 DLL is nowhere in the path, OCaml executables just crash without any explanation. Will our Windows users understand this? (They are used to suffer already...) Could pthread-win32 be linked statically instead?

I've experimented with this manually (by editing `Makefile.config`) and it seems fine. I only need to wrestle with flexlink (which might need changing, as you either have to pass the full path to the static library, or enhance flexlink's interpretation of `-lpthreads`).

File: otherlibs/systhreads/st_posix.h (hunk 1 above)
Comment (xavierleroy) [reply otherlibs/systhreads/st_posix.h]: Yes, I'd rather leave out the last commit (the one that supports both select and nanosleep).  There is no strong reason to prefer one to the other!  It's just the comment about "the most efficient way" that I'd like to see removed.  But let's not spend too much time on this.

File: otherlibs/systhreads/st_stubs.c (hunk 3 above)
Comment (xavierleroy) [reply otherlibs/systhreads/st_stubs.c:37]: I agree it's enough to include `../../runtime/sync_posix.h` once in `st_stubs.c`.  Hardening `sync_posix.h` against multiple inclusions doesn't hurt and could be useful later on.

File: otherlibs/systhreads/st_posix.h (hunk 1 above)
Comment (dra27) [reply otherlibs/systhreads/st_posix.h]: Done - I've parked that commit as a reminder in #10505.

File: otherlibs/systhreads/st_stubs.c (hunk 3 above)
Comment (dra27) [reply otherlibs/systhreads/st_stubs.c:37]: OK - second include deleted, but the guard in `runtime/sync_posix.h` left.

**Comment by dra27** [comment]
History cleaned up - the `nanosleep` change is removed and the original comment excised in https://github.com/ocaml/ocaml/pull/11115/commits/6d1beb3a2e0bf7d5c014302ea4ecfaf2b6ffd6c4 and `sync_posix.h` is now only `#include`'d by `st_stubs.c`. I'm just running it through [precheck#668](https://ci.inria.fr/ocaml/job/precheck/668/), but assuming all the CI agrees, I think this is good to go?

**Comment by dra27** [comment]
Just to ensure I record it somewhere, I looked at the wall clock time for all the tests on Windows which use systhreads using a GitHub Actions runner. I didn't do enough runs to eliminate noise for most of the tests (two runs with this PR, one had _all_ tests slightly slower with this PR and one had _all_ tests slightly faster, so noise), however three tests stick out consistently:

- lib-threads/sieve halved its runtime with this PR
- parallel/fib_threads slightly more than halved its runtime with this PR
- parallel/multicore_systhreads nearly halved its runtime with this PR

In Dune's case, it was the difference between a small library taking _5 minutes_ to build vs a matter of seconds.

| test | before/s | run1/s | run2/s |
| ---- | --------| ----- | ----- |
| backtrace/backtrace_systhreads | 8.82 | 9.25 | 8.55 |
| backtrace/callstack | 2.61 | 3.13 | 2.55 |
| lib-channels/input_all | 3.39 | 3.43 | 2.84 |
| lib-dynlink-native/main | 40.63 | 51.54 | 39.60 |
| lib-systhreads/multicore_lifecycle | 12.42 | 13.24 | 12.34 |
| lib-systhreads/test_c_thread_register | 6.51 | 7.77 | 5.70 |
| lib-threads/backtrace_threads | 3.28 | 3.17 | 2.27 |
| lib-threads/bank | 3.43 | 3.11 | 2.20 |
| lib-threads/bufchan | 3.19 | 3.13 | 2.34 |
| lib-threads/close | 5.30 | 5.09 | 4.32 |
| lib-threads/fileio | 4.37 | 4.71 | 3.20 |
| lib-threads/mutex_errors | 2.51 | 3.13 | 2.34 |
| lib-threads/pr4466 | 3.04 | 3.58 | 2.75 |
| lib-threads/pr5325 | 4.45 | 4.85 | 4.33 |
| lib-threads/pr7638 | 2.38 | 3.20 | 2.30 |
| lib-threads/pr8857 | 2.25 | 3.30 | 2.36 |
| lib-threads/prodcons | 2.73 | 3.09 | 2.36 |
| lib-threads/prodcons2 | 3.57 | 3.58 | 2.88 |
| **lib-threads/sieve** | **5.36** | **3.18** | **2.47** |
| lib-threads/swapchan | 2.56 | 3.10 | 2.42 |
| lib-threads/tls | 2.85 | 3.36 | 2.42 |
| lib-threads/torture | 9.39 | 9.14 | 8.39 |
| lib-threads/uncaught_exception_handler | 2.53 | 3.30 | 2.35 |
| lib-unix/win-socketpair/test.ml | 2.51 | 3.27 | 2.37 |
| **parallel/fib_threads** | **137.52** | **62.29** | **63.95** |
| **parallel/multicore_systhreads** | **20.20** | **11.88** | **11.08** |
| parallel/prodcons_domains | 2.83 | 3.78 | 2.75 |
| parallel/test_c_thread_register | 7.13 | 8.09 | 6.86 |

**Review by xavierleroy — APPROVED** [review]
Looks good to me, thanks a lot!
